Allow digits in the `NSID`; fix edge cases for string format validation

zzstoatzz opened the pull request:
see https://github.com/bluesky-social/atproto-website/pull/402


and https://bsky.app/profile/bnewbold.net/post/3lk5euaqmic2o

zzstoatzz commented on tests/test_atproto_client/models/tests/test_string_formats.py:
```diff
@@ -202,3 +202,28 @@ class FooModel(BaseModel):
     assert 'Invalid DID' in str(exc_info.value)
+
+
+def test_nsid_with_digits_in_final_segment() -> None:
+    """Test that NSIDs with digits in the final segment are now valid."""
+    test_cases = [
+        'com.example.postV2',
+        'app.bsky.feed.post123',
+        'com.example.v2',
+        'net.users.bob.post99',
+    ]
+
+    nsid_adapter = TypeAdapter(string_formats.Nsid)
+    for nsid in test_cases:
+        # Should pass with strict validation
+        validated = nsid_adapter.validate_python(nsid, context={_OPT_IN_KEY: True})
+        assert validated == nsid
+
+    # These should still be invalid (digits in non-final segments)
+    invalid_cases = [
+        'com.example123.post',
+        'app.bsky2.feed.post',
+    ]
+    for invalid_nsid in invalid_cases:
+        with pytest.raises(ValidationError):
+            nsid_adapter.validate_python(invalid_nsid, context={_OPT_IN_KEY: True})
```
since the interop files were updated, strictly speaking we shouldnt need this I think - happy to remove

MarshalX replied:
yeah, i think interop files will cover this nicely

MarshalX commented:
thank you! Should we update this as well? the regex https://atproto.blue/en/latest/atproto_core/nsid.html

also, it is worth to validate manually the logic of code generators; i do not expect any problems with numbers here since they are not leading but just in case

zzstoatzz commented:
hi @MarshalX - sorry for the delay but coming back to this I think I've addressed what you mentioned, though let me know if I misinterpreted

> it is worth to validate manually the logic of code generators

I've also removed the redundant (because of the interop files) unit tests

MarshalX commented:
@zzstoatzz i am so sorry for such a long review. i suspect something wrong in the implementation. that's why i connected interop test files to NSID class testing. And as we can see tests are failed.

The problem that I suspect is that even segments in the middle could use digits without any problem. And in the code i can see that this is forbidden. 

The good news is that I aligned NSID validation with the official version and now all tests passed! I will push it in the next commit!

MarshalX commented:
@zzstoatzz I found out that `test_string_format_validation` test valid cases against `valid_data` fixture. Which is only the first valid example from the file. So I fixed that to use all valid examples from interop files. We have 53 failed tests :( affected types are: AtUri, datetime, handle, did and nsid. Could you please take a look?

MarshalX commented:
@zzstoatzz let me know if you want to separate these fixes into another PR. But let's fix at least NSID here

zzstoatzz commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -193,13 +206,7 @@ def validate_nsid(v: str, _: ValidationInfo) -> str:
     Raises:
         ValueError: If NSID format is invalid
     """
-    if (
-        not atproto_core_validate_nsid(v, soft_fail=True)
-        or len(v) > MAX_NSID_LENGTH
-        or any(c in v for c in '@_*#!')  # Explicitly disallow special chars
-        or any(len(seg) > 63 for seg in v.split('.'))  # Max segment length
-        or any(seg[-1].isdigit() for seg in v.split('.'))  # No segments ending in numbers
-    ):
+    if not atproto_core_validate_nsid(v, soft_fail=True):
```
i was causing failures by incorrectly reapplying bits of logic that now live in the core validation

zzstoatzz commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -343,35 +352,65 @@ def validate_datetime(v: str, _: ValidationInfo) -> str:
+        # Handle both Z and explicit timezone formats
+        datetime_str = v
+        if v.endswith('Z'):
+            # For Python 3.10 compatibility, normalize fractional seconds to 6 digits
+            if '.' in v:
+                base, frac = v[:-1].split('.')
+                # Pad or truncate to exactly 6 digits
+                frac = (frac + '000000')[:6]
+                datetime_str = f'{base}.{frac}+00:00'
+            else:
+                datetime_str = v[:-1] + '+00:00'
```
it seems that 3.10 has stricter requirements on fractional seconds, so I'm normalizing to 6 digits

MarshalX replied:
do you know the link to Python`s changelog about it or something? would be great to mention it in the comment

zzstoatzz replied:
added in [1b31442](https://github.com/MarshalX/atproto/pull/572/commits/1b31442342cc981dcae7f6ac8046476d4bd3448d)

zzstoatzz commented:
hi @MarshalX

> @zzstoatzz I found out that test_string_format_validation test valid cases against valid_data fixture. Which is only the first valid example from the file

thanks for catching that! my mistake 😅 

While keeping your updates to the test collection, I've updated the validation for nsids, at-uris, dids, handles, and datetimes and things seem to be passing now. Most egregiously it seems I had been doing some naive timezone parsing for datetime (that was failing hard against some of the interop cases) which should now be fixed. Importantly, I also "fixed" the nsid validator in `string_formats.py` simply by removing extra logic and deferring entirely to your new nsid validation introduced by [20d45c5](https://github.com/MarshalX/atproto/pull/572/commits/20d45c5b98665cdd85a94786adb7f2dcb379869e).

I tried to stage these fixes out by commit in case you decide we ought to separate some of this out into another PR, but it didn't end up sprawling too much so I figured we could try it all in one go 🙂 

let me know what you think!

MarshalX commented:
@zzstoatzz that's great! we can merge all fixes in this PR, no problem

MarshalX commented on packages/atproto_client/models/string_formats.py:
```diff
@@ -193,13 +206,7 @@ def validate_nsid(v: str, _: ValidationInfo) -> str:
-        or any(len(seg) > 63 for seg in v.split('.'))  # Max segment length
-        or any(seg[-1].isdigit() for seg in v.split('.'))  # No segments ending in numbers
-    ):
+    if not atproto_core_validate_nsid(v, soft_fail=True):
         raise ValueError(
             f'Invalid NSID: must be dot-separated segments (e.g. app.bsky.feed.post) with max length {MAX_NSID_LENGTH}'
         )
```
```suggestion
    try:
        atproto_core_validate_nsid(v)
        return v
    except InvalidNsidError as e:
        raise ValueError(f'Invalid NSID: {e!s}') from None
```

let's proxy error messages here; and delete `MAX_NSID_LENGTH` const

zzstoatzz replied:
ah, good call! updated in [22aff21](https://github.com/MarshalX/atproto/pull/572/commits/22aff2117575fd35560f3bdcc626baf872e6163d)

MarshalX commented:
i am ready to merge. let me know (like reaction is enough) that this is done

zzstoatzz commented:
> i am ready to merge. let me know (like reaction is enough) that this is done

![image](https://github.com/user-attachments/assets/4b458429-5e8b-4f43-8eb2-67edd95a25c4)
